libgadget/lenstools.c

```diff
@@ -0,0 +1,412 @@
+    // Get the rank of the current process
+    int ThisTask;
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    // smooth
+    float smooth = 1.0; // fixed in our case
+
+    int num_particles_rank = PartManager->NumPart;  // dark matter-only simulation: NumPart = number of dark matter particles
+
+    double **density_projected = allocate_2d_array(plane_resolution, plane_resolution);
+
+    // printf("cutPlaneGaussianGrid called with rank %d\n", rank);
+    // int i, j;
+    int thickness_resolution = 1;  // Number of bins along the thickness direction, fixed to 1 for now
+
+    // cosmological normalization factor
+    double H0 = 100 * CP->HubbleParam * 3.2407793e-20;  // Hubble constant in cgs units
+    double cosmo_normalization = 1.5 * pow(H0, 2) * CP->Omega0 / pow(LIGHTCGS, 2);  
+
+    // Binning for directions perpendicular to 'normal'
+    double *binning[3];
+    int plane_directions[2] = { (normal + 1) % 3, (normal + 2) % 3 };
+
+    for (int i = 0; i < 3; i++) {
+        int resolution = (i == normal) ? thickness_resolution : plane_resolution;
+        binning[i] = malloc((resolution + 1) * sizeof(double));
+        if (binning[i] == NULL) {
+            fprintf(stderr, "Memory allocation failed for binning %d on rank %d\n", i, ThisTask);
+            // Free other allocated resources
+            for (int j = 0; j < i; j++) {
+                free(binning[j]);
+            }
+            MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
+        }
+        double start = (i == normal) ? (center - thickness / 2) : left_corner[i];
+        double stop = (i == normal) ? (center + thickness / 2) : (left_corner[i] + Lbox);
+        linspace(start, stop, resolution + 1, binning[i]);
+    }
+
+
+    // bin resolution (cell size in kpc/h)
+    double bin_resolution[3];
+    bin_resolution[plane_directions[0]] = Lbox / plane_resolution;
+    bin_resolution[plane_directions[1]] = Lbox / plane_resolution;
+    bin_resolution[normal] = thickness / thickness_resolution;
+
+    // density normalization
+    double density_normalization = bin_resolution[normal] * comoving_distance * pow(CM_PER_KPC/CP->HubbleParam, 2) / atime;
+
+
+    // density 3D array
+    GridDimensions dims;
+    // Set dimensions based on the orientation specified by 'normal'
+    dims.nx = (normal == 0) ? thickness_resolution : plane_resolution;
+    dims.ny = (normal == 1) ? thickness_resolution : plane_resolution;
+    dims.nz = (normal == 2) ? thickness_resolution : plane_resolution;
+    // printf("nx: %d, ny: %d, nz: %d\n", dims.nx, dims.ny, dims.nz);
+
+    double ***density = allocate_3d_array(dims.nx, dims.ny, dims.nz);
+
+    grid3d_nfw(P, num_particles_rank, binning, dims, density);
+
+    projectDensity(density, density_projected, dims, normal);
```

Comment:
Am I reading this correctly that it is only density_projected that is used below? Can we skip computing density completely and save the memory?

Reply:
Yes, currently only density_projected is used, but I recommend reserving the 3D density since it could be useful in future (3D density planes for visualization for example), and the density should not occupy much memory since we now have fixed thickness_resolution to 1 (in future if needed, we can add ThicknessResolution and PlaneKind to generate 3D density planes):

`int thickness_resolution = 1;  // Number of bins along the thickness direction, fixed to 1 for now`

density and density_projected should be equal in size and they are not very large, typically e.g., 2048^2 double (64 MB), what do you think?

Reply:
We can already compute the 3D density in petapm.c very efficiently. That code is quite involved and complicated because of the MPI parallelisation: each processor has a different part of physical space. The code is set up so that it only allocates memory for that specific part of physical space. For these arrays if they are really that small it makes sense to allocate them on each processor and reduce them to make the code simpler.

If I understand right though, density_projected doesn't do anything when thickness_resolution is 1? So we don't need two arrays, I think. Even if you wanted to keep the 3D array it would be more efficient to reduce it to 2D in place.